KEP-4188: New kubelet gRPC API with endpoint returning local pods information

kl52752 opened the pull request:
<!-- 
	Please use the following format when naming your PR
	< Issue Number >:< Issue Description >
	e.g. KEP-000: adding beta graduation criteria
	
	Avoid using phrases like `fixes #NNNN` in the description
	unless the pull request is to change the KEP status to 
	implemented or KEP has been deprecated.
-->

<!-- short description of work done in PR e.g. updating milestone, adding new KEP, adding test requirements… -->  
- One-line PR description: Proposal for new kubelet gRPC API with endpoint returning local pods information

<!-- link to the k/enhancements issue -->
- Issue link: https://github.com/kubernetes/enhancements/issues/4188

<!-- other comments or additional information -->
- Other comments:

kl52752 commented:
/cc @robscott 
/assign @SergeyKanzhelev

ffromani commented:
/cc

kl52752 commented:
> We actually do expose pod readiness information via metrics: https://github.com/kubernetes/kubernetes/blob/e00af59d0fccf7a0219bb93218c06281a81787f8/pkg/kubelet/prober/prober_manager.go#L38-L65

Thanks for pointing that out. I can see that this metrics is only exported for pod's that actually defined the probes but with this API what we want to achieve is the readiness status of all pods running on the node. Readiness information is computed based on different Conditions and Probes are just one of them.

ffromani commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
````diff
@@ -0,0 +1,343 @@
+```protobuf
+service PodInfo {
+    // ListPodConditions returns a stream of List of PodConditions
+    rpc ListPodConditions(Empty) returns (stream PodConditionsResponse) {}
````
`List` does not _usually_ returns a stream, `ListAndWatch` or `Watch` do. Do we need a stream here or just a simple response?

SergeyKanzhelev replied:
sorry, it does still returns a stream.

ffromani commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,343 @@
+
+###### Will enabling / using this feature result in introducing new API types?
+
+Yes. We are adding new proto messages to be use with this API.
```
I think this refers to HTTP public-facing APIs, but we may be in a grey area here

ffromani commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,343 @@
+
+### Non-Goals
+
+### User Stories
+* Some on node system workloads want to reduce Control Plane dependency and introduce locality for pod’s readiness due to reliability and scalability issues.
+* For building custom monitoring tools that show information about pod’s readiness. When it is ready to serve the traffic.
```
I for myself I'll need to review the past conversations about the topic which KEP wants to cover. This is my biggest concern. If exposing this data is ok at architecture (sig-node/sig-arch) level, everything else I see here seems pretty sensible straightforward to me, pending more review cycles.
I'm not sure about the consequences of tolerating control plane unavailabvility and effectively duplicate this information.

robscott replied:
Yep, we've been discussing this proposal at both sig-node and sig-arch meetings (this is on the agenda for tomorrow's sig-arch meeting if you'd like to join).

From my biased perspective, I think it's important to remove unnecessary dependencies on the control plane wherever we can. In the case of health checking in particular, this information is originating from Kubelet, so a node-local component fetching this from Kubelet would have 2 advantages:

* Reduced load on control plane
* Improved reliability if/when control plane is unavailable

ffromani commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,343 @@
+
+#### Control Plane availability issue
+
+The Kubernetes Control Plane is the source of truth and guarantees system integrity. However, already running pods should not be dependent on Control Plane availability or connectivity issues. When Node is healthy but has problems accessing Kube-API running pods, statuses should be refreshed and stored as usual.
+The new API should return pods statuses that are computed by kubelet even though they are not reported back to kube-API but reflect the current pod’s state.
```
in a nutshell, this is what I'm kinda concerned: the control plane is the source of truth... but in some cases the kubelet-provided data will be the actual source of truth. Perhaps a step in the right direction would be to make sure that the information is provided by this API only if it is acknowledged by the control plane somehow (e.g. after a succesfull update? that would require firther thought) and so no information misalignment is ever possible?

robscott replied:
The actual purpose of this GEP is to remove a dependency on control plane availability here. This is coming from a networking perspective where we're trying to understand the health of Pods running locally on the same Node. Since the readiness probes are run locally with Kubelet, it does not make sense for us to have a dependency on control plane for this information. 

Essentially we're trying to avoid a case where we're not updating our routing logic if the control plane is unavailable. This could cause completely unnecessary downtime for applications running on Kubernetes.

dchen1107 replied:
@ffromani I think @robscott's reply should address your concerns above. Basically the proposed API won't be used to override any decision made by the control plane.

ffromani commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,343 @@
+
+#### GA
+
+- [ ] No issue related to usage of API.
```
you may likely want to have a bit more precise criterias here (e.g. no major bugs in X months/cycles, or anything you see fit). May be fixed later.

ffromani commented on keps/sig-node/4188-kubelet-pod-readiness-api/kep.yaml:
```diff
@@ -0,0 +1,43 @@
+# The milestone at which this feature was, or is targeted to be, at each stage.
+milestone:
+  alpha: "v1.29"
+  beta: "v1.30"
+  stable: "v1.31"
```
this seems a very aggressive schedule, but can be reviewed later and adjusted if needed

ffromani replied:
In general it's a good practice to give some time (like a couple of release cycles) to folks to consume and integrate the change, so bugs or issues or rough edges surface. But we can surely review later and discuss during the KEP graduation process.

ffromani reviewed (commented):
non-binding-my-2c-initial-review: most of what we read here seems sensible and straightforward. But the elephant in the room IMO is: do we actually _want_ to provide this information straight from the kubelet and tolerate control plane unavailability? I'm worried about information misalignment (a case of split-brain).
The answer may very well be "yes" because I missed the initial conversation with lead to the KEP (catching up); in that case, I think it should be more explicitely conveyed in the KEP.

robscott commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,343 @@
+
+### Goals
+
+The goal of this API is to expose Pod readiness information directly from the source - kubelet. This would remove the need for node-local components to request this node-local information from the Kubernetes API Server. Kubelet already has a podresources endpoint ([2403-pod-resources-allocatable-resources](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2403-pod-resources-all ...
```
These endpoints are not intending to be consistent. Each endpoint would represent the state of the world from the perspective of that specific component.

ffromani replied:
fair enough

ffromani commented:
@kl52752 please add the `production readiness review` [metadatata](https://github.com/kubernetes/community/blob/master/sig-architecture/production-readiness.md)

kl52752 commented:
> metadatata

@ffromani can you be more specific? I filled the section `Production Readiness Review Questionnaire` in README.md?
Do you want me to follow stepps to submit the KEP for review? https://github.com/kubernetes/community/blob/master/sig-architecture/production-readiness.md#submitting-a-kep-for-production-readiness-approval

ffromani commented:
> > metadatata
> 
> @ffromani can you be more specific? I filled the section `Production Readiness Review Questionnaire` in README.md? Do you want me to follow stepps to submit the KEP for review? https://github.com/kubernetes/community/blob/master/sig-architecture/production-readiness.md#submitting-a-kep-for-production-readiness-approval

Sure: your PR needs to add a file like `prod-readiness/sig-node/4188.yaml` whose content shall include the Production Readiness review, so the reviewers can add to their review lists. Example: https://github.com/kubernetes/enhancements/blob/master/keps/prod-readiness/sig-node/3570.yaml
If you don't know which Prod Readiness Reviewer put in the PRR metadata, pick any: the team will rebalance internally and suggest the right name in due time.

kl52752 commented:
> Sure: your PR needs to add a file like `prod-readiness/sig-node/4188.yaml` whose content shall include the Production Readiness review, so the reviewers can add to their review lists. Example: https://github.com/kubernetes/enhancements/blob/master/keps/prod-readiness/sig-node/3570.yaml If you don't know which Prod Readiness Reviewer put in the PRR metadata, pick any: the team will rebalance internally and suggest the right name in due time.

Thanks, added the file

ffromani commented on keps/prod-readiness/sig-node/4188.yml:
```diff
@@ -0,0 +1,7 @@
+kep-number: 4188
+alpha:
+  approver: "@dchen1107"
```
I'm not sure Dawn is among PRRs
In general, we just add the approver on stage by stage basis. In your case you just need to add `alpha`

kl52752 replied:
I chose here from [OWNERS_ALIASES](https://git.k8s.io/enhancements/OWNERS_ALIASES) from sig-node-tech list.

ffromani replied:
Sure! For the sake of the awarness, production readiness reviewers are by design a disjoing set from the partecipating SIGs. Is meant to be a separate review from different people.

aojea replied:
Wojtek, David Eads or John Belamaric. I think there is new people these days also, but those are the top level approvers

kl52752 replied:
assigned to @johnbelamaric

kl52752 replied:
@johnbelamaric is that ok for you to review the PRR for this KEP?

swatisehgal commented:
/cc

MarSik commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,345 @@
+| --------------------------------------------------------- | ------------- | ---------- |
+| Too many requests to the API impacting the kubelet performances | High          | Rate limiting the API. |
+| Misuse of the API  | High |  This API is Read-only. We will expose only a small portion of the pod's information related to the pod readiness. Exposed data does not contain sensitive information that could be used in a malformed way. |
+| Unauthorized access to the API | Medium | This API is designed to be accessed by all on-node workloads. There is no need to authorize the clients to access this API. |
```
Why not? The API server RBAC rules allow isolating namespaces for example. How will this work in a multi-tenant environments? Will all potential users of a cluster have read access to pod and container names, statuses etc of all the other customers' workloads?

kl52752 replied:
yes, this API is design to work for system pods that needs to know the state of the node (all running workloads). Adding API server RBAC rules will introduce ControlPlane dependency on this API and the reason why we propose this API is that we want to fetch the data directly form kubelet without ControlPlane dependency. 
If someone wants the isolation then Kube-API should be used and this API can be enabled/disabled with feature gate.

ffromani replied:
>If someone wants the isolation then Kube-API should be used and this API can be enabled/disabled with feature gate.

dchen1107 replied:
@ffromani good point! Since the API is read-only, I don't have the strong opinion on reading other's pod/container name and status, but we might want to consult with SIG Auth folks. I won't treat this as the blocker issue for alpha though.

robscott replied:
For reference, I believe both @deads2k and @liggitt provided feedback on this KEP when we met with sig-arch about this proposal on Aug 24 and Sep 7. I may be misremembering here, but I'd thought that at least Pod readiness was considered acceptable to expose to workloads running on the same Node.

soltysh replied:
> that at least Pod readiness was considered acceptable to expose to workloads running on the same Node.

I'd like to make sure this KEP only touches the pod readiness. At least suggests more information is considered, and if so it should be explicitly specified to ensure we don't leak sensitive information

MarSik commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,345 @@
+* The API will return data about both Static and Regular Pods.
+* The API will not return partial data. If kubelet does not know actual information about workloads then appropriate error code will be returned.
+* API should return the most recent information about pods computed by the kubelet even when those data were not reported or accepted by kube-API.
+* The API will be read-only and accessible for on-node workloads (we will use unix socket to connect) without the authorization.
```
Do you have an example workload spec? How will you expose the unix domain socket to the Pods?

kl52752 replied:
in your client's pod spec you need to mount the path to the socket file in container. 

Just like for pod-resource API:
```
volumeMounts:
        - mountPath: /var/lib/kubelet/pod-resources/kubelet.sock
          readOnly: true
          name: podresourcesocket
      volumes:
      - name: podresourcesocket
        hostPath:
          path: /var/lib/kubelet/pod-resources/kubelet.sock
          type: Socket
```

This new API will have different path :)

swatisehgal reviewed (commented):
Have you considered adding podreadiness endpoints to the existing podresource API rather than introducing a new Kubelet gRPC API?

Being able to obtain pod readiness information prior to querying for allocated resources seems like a nice enhancement to the existing pod resource API and helps us achieve the goals of this KEP. It would be nicer from end user/client perspective to have a unified API for node local information rather than having to access to multiple unix sockets in case both these capabilities are required.

kl52752 commented:
> Have you considered adding podreadiness endpoints to the existing podresource API rather than introducing a new Kubelet gRPC API?
> 
> Being able to obtain pod readiness information prior to querying for allocated resources seems like a nice enhancement to the existing pod resource API and helps us achieve the goals of this KEP. It would be nicer from end user/client perspective to have a unified API for node local information rather than having to access to multiple unix sockets in case both these capabilities are required.

hi, because our goal with this API is to reduce the Control Plane dependency we didn't wanted to add this to already existing API.

ffromani commented:
> > Have you considered adding podreadiness endpoints to the existing podresource API rather than introducing a new Kubelet gRPC API?
> > Being able to obtain pod readiness information prior to querying for allocated resources seems like a nice enhancement to the existing pod resource API and helps us achieve the goals of this KEP. It would be nicer from end user/client perspective to have a unified API for node local information rather than having to access to multiple unix sockets in case both these capabilities are required.
> 
> hi, because our goal with this API is to reduce the Control Plane dependency we didn't wanted to add this to already existing API.

This is another area I for myself don't have strong opinions about. While having a consistent snapshot is very nice, there are arguments for both directions. Could you maybe add a section with pros/cons in the KEP design details? Especially considering that also podresources API has no control plane dependency.

kl52752 commented:
>  Could you maybe add a section with pros/cons in the KEP design details? Especially considering that also podresources API has no control plane dependency.

you mean the pros and cons of the API? I've already added the table in Motivation section the pros and cons  of the API itself and also there I explain why we want this API to be control plane independency.
If sth is missing there I will add the details.

I just wanted to highlight that the goal of this API is to be control plane independent (I've updated `goal` section).

Regarding podresource API - I'm open to discuss adding this API to already existing one podresource if it will not introduce the needs for Authorization (kube-API dependency :))

kl52752 commented:
@ffromani  @swatisehgal I added new tables where I gather the information about options for the API. One is gRPC vs REST and another one the option to integrate with podresource api.
https://github.com/kubernetes/enhancements/pull/4184/files#diff-f8afbc8d85850ed206e915bb96da86ccbfa2cd572292e6e0638c84ca7f122fcaR109

I didn't add this to summary but I can see 2 options for this API:|
1. use REST API - for better adoption 
2. use gRPC and integrate with pod resource API - I'm not sure if this is desired direction 
@mrunalp @dchen1107

robscott commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,365 @@
+| Authorization | none, using unix socket to limit the clients to local workloads | none, using unix socket to limit the clients to local workloads |
+| access | Read-Only | Read-Only|
+
+Podresource API looks like a good fit for this endpoint. Our major concern is how will this integration impact the shape of this API in the future.
```
I think I'd also note that we expect there to be little-to-no overlap in the clients for both of these APIs, they're both serving information that would interest different sets of clients. Admittedly this may converge over time as new features are requested, but it may make sense to keep these separate so clients don't need to fetch data they don't need.

kl52752 replied:
done

robscott commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,365 @@
+
+### Proposed API
+
+We propose to add new gPRC API `podinfo` in Kubelet, listening on a unix socket at `/var/lib/kubelet/pod-info/kubelet.sock`. The endpoint will be versioned.
```
Maybe instead of referring to this as `podInfo` we should consider it `podStatus` or similar? Not sure on the name here, but `podInfo` is probably broader than we'd want here.

kl52752 replied:
podInfo is more flexible and we don't limit this API shape in this early phase.

robscott reviewed (commented):
Thanks @kl52752!

kl52752 commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,365 @@
+## Motivation
+
+Kubelet is responsible for running Health Checks (probes) and communicating the results via Pod status. All that information is stored in cache and reported to Kube-API. Right now pod information is tightly coupled with the Kubernetes API Server. When a workload wants to know the actual state of pods running on the Node, it needs to fetch it from Kube-API.
+This causes some issues:
+Reliability - for various reasons Kube-API might not be available (connectivity issue, control plane updates)
+Scalability - adding new watchers to kube-API is a scalability concern
+By exposing the endpoint that will serve the pods status directly from the kubelet cache we can use  it on node workloads and avoid additional dependencies on the Kubernetes API Server.
```
fixed

kl52752 commented on keps/prod-readiness/sig-node/4188.yaml:
```diff
@@ -0,0 +1,7 @@
+beta:
+  approver: "@johnbelamaric"
+stable:
+  approver: "@johnbelamaric"
```
this is required by the check Error:      	"map[/home/prow/go/src/github.com/kubernetes/enhancements/keps/sig-node/4188-kubelet-pod-readiness-api/kep.yaml:[getting PRR approver for alpha stage: none of the PRR milestones are populated]]" should have 0 item(s), but has 1

I tried to remove the approver and left it empty but it still fails.

kl52752 replied:
oh yeah I removed everything :D

robscott commented:
I think this is ready for PRR review. @johnbelamaric recommended asking for a shadow PRR review first, @soltysh and/or @BenTheElder could you help with that?

/cc @soltysh @BenTheElder

dchen1107 commented:
Had several offline review with the team, and the current KEP addressed my concerns: 1) Read-only 2) throttling  3) limited to small set of information initially, for example, pod condition. It also answered why not use PodResource API, which was raised during SIG Node weekly. 

/lgtm I will wait for @mrunalp and other maintainers to chime in before finally approve it.

soltysh commented:
> I think this is ready for PRR review. @johnbelamaric recommended asking for a shadow PRR review first, @soltysh and/or @BenTheElder could you help with that?
> 
> /cc @soltysh @BenTheElder

looking at it...

soltysh commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,366 @@
+
+#### Control Plane availability issue
+
+The Kubernetes Control Plane is the source of truth and guarantees system integrity. However, already running pods should not be dependent on Control Plane availability or connectivity issues. When Node is healthy but has problems accessing Kube-API running pods, statuses should be refreshed and stored as usual.
```
kube-apiserver not accessible from a given node isn't the sign of its unavailability. It's possible that the node lost its network connectivity and the rest of the system will eventually consider the node down. So this assumptions is only partially valid, in very narrow case, which I'd like to be explicitly laid out here.

robscott replied:
Fair point, added some additional content to clarify this. A high level summary is that since this would only impact node-local components, and the Node would essentially be cut off from the rest of the cluster in this state, it should not be a concern for this API.

soltysh commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,366 @@
+#### Alpha
+
+- [ ] Feature implemented behind a feature flag.
+- [ ] e2e tests completed and enabled.
```
sig-auth input about RBAC is a must

robscott replied:
Added as part of graduation criteria.

soltysh reviewed (commented):
This KEP should not move forward without discussing with sig-auth the potential risks from exposing pod information without proper RBAC.

robscott commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,366 @@
+| + | Safety - Read-only API will not add security risks. |
+| + | Reduce resource consumption by workload. Using Kube-API we can fetch objects like PodSpec or PodStatus, for some on-node workload this might be unnecessary, with this API workload can reduce the resource consumption. |
+| - | This API will add load to kubelet (mitigation: API will be rate limited) |
+| - | Kubelet does not support RBAC authorization for gRPC. (mitigation: This API is design to be accessible for all workloads running on the node without the authorization. The unix socket will be used for the connection and all exposed data will be carefully reviewed.) |
```
We have discussed that at sig-arch previously, which had a fairly good representation from sig-auth. The idea is that what we're doing here is not particularly novel. We already have a Pod resources API in Kubelet that is exposed in a similar way (via socket and therefore only accessible to other workloads on the same Node). We believe that Pod readiness is safe to make available to workloads running on the same node. I think both @deads2k and @liggitt have some context here, maybe they can chime in here. 

In the future the there may be a proposal to increase the scope of the information exposed by this endpoint, but that should go through a separate audit before it is approved.

soltysh replied:
If this is about readiness only that is acceptable, the problem I struggled with, was that in multiple places various pieces of information were alluding to more broadly exposed information:
1. in summary: `Proposal to add a new Kubelet gRPC API with endpoint returning local Pods information.`
2. in motivation section, there are multiple statements alluding to `local Pods data`
3. in goals section: `create a new API that will expose pod conditions that are computed` 
4. in risks and mitigations: `We will expose only a small portion of the pod's information related to the pod readiness.` 
5. also in risks and mitigations: `his API should serve only complete information about workloads. If Kubelet is in the init phase and not all pod’s information is known, then the API should report that information.` 
6. in control plane availability issue: `The new API should return Pods statuses that are computed by Kubelet even if
they are not reported back to kube-API but reflect the current Pod’s state locally.`
7. the API proposal calls method `ListPodConditions`, `WatchPodConditions`, `GetPodConditions`

I'd suggest rewording those places, to ensure that it's explicitly stated the scope of the expose information. My current confusion came from the fact that various places mentioned various scopes of exposed data, once it was called conditions, other places - pod status, etc. Where each of that information seems to me more than just pod readiness you're talking about.

SergeyKanzhelev replied:
Good point. Let's just make sure that we explicitly mention that future APIs expansion is expected. This API is limited to pod readiness in this KEP, but open for future KEPs. For example, we wouldn't name the socket `pod-readiness` and will use way more generic name.

kl52752 replied:
@soltysh @SergeyKanzhelev I updated the doc and changed all those places to point out that we want to expose readiness information. Regarding the methods name I will leave it as Conditions since at this phase the Pod's condition form v1.PodStatus will be exposed via this API. 
With this API we don't want to just expose the boolean value of Ready status (True or False) but more granular state of the pod's readiness which is reflected in those Conditions.

soltysh replied:
For API methods it's fine to have them built with expansion in mind, but as long as the rest of the document, and the comments on those method are explicit about what functionality is available it should be all good.

dchen1107 replied:
+1 and will take another pass.

robscott commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,366 @@
+## Proposal
+
+We are proposing to create a new kubelet API that will return pod's readiness information.
+* The API will return data about both Static and Regular Pods.
```
We don't believe that Pod readiness information is sensitive. Do you have a specific form of escalation in mind here?

robscott commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,366 @@
+We are proposing to create a new kubelet API that will return pod's readiness information.
+* The API will return data about both Static and Regular Pods.
+* The API will not return partial data. If kubelet does not know actual information about workloads then appropriate error code will be returned.
+* API should return the most recent information about pods computed by the kubelet even when those data were not reported or accepted by kube-API.
```
The fundamental purpose of this KEP is to ensure that there is a path to get Pod readiness information without a dependency on the Kubernetes control plane. One of the great parts about Kubernetes is that it can coast reasonably well if the control plane is unavailable. One problem is that if Pod readiness ever changes in that state, networking components are going to be completely unaware of that because it has a control plane dependency. Removing that dependency as proposed here may improve both the reliability and performance of the system.

SergeyKanzhelev commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,366 @@
+
+### Proposed API
+
+We propose to add new gPRC API `podinfo` in Kubelet, listening on a unix socket at `/var/lib/kubelet/pod-info/kubelet.sock`. The endpoint will be versioned.
```
If in future we will want to return information about config maps, I think we can use the same API. What about `/var/lib/kubelet/status/kubelet.sock`? Or some similar more generic name?

kl52752 replied:
you mean to change the API name to status not PodInfo?

SergeyKanzhelev commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,366 @@
+
+// PodConditionsResponse returns a stream of List of PodConditionsResponse
+// Whenever a Pod state changes it will return the new list.
+message PodConditionsListResponse {
```
so what does API return when kubelet is not yet fully initialized?

SergeyKanzhelev replied:
Do we need some status field here?

kl52752 replied:
no I think we should return gRPC error codes. https://grpc.github.io/grpc/core/md_doc_statuscodes.html
When kubelet is not initialized we will return FAILED_PRECONDITION(9) error code.

SergeyKanzhelev commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,366 @@
+    // Whenever a pod state change api returns the new list
+    rpc WatchPodConditions(Empty) returns (stream PodConditionsResponse) {}
+    // GetPodConditions returns a PodInfo for given pod's UID
+    rpc GetPodConditions(GetPodConditionsRequest) returns (PodInfo) {}
```
let's not call the type the same as a service

SergeyKanzhelev commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,366 @@
+}
+
+// PodInfo returns a Pod details and list of status Conditions
+message PodInfo {
```
PodStatus has last message and last reason as a top level values. Do you want to do the same?

kl52752 replied:
no I don't think it is needed.

SergeyKanzhelev replied:
so the way to get those would be to check sort conditions by timestamp? Or you are saying this information is not needed?

kl52752 replied:
I think it is not needed. We have max 5 conditions and every condition occurs once in the list so based on that condition the workload can compute the readiness status.
I other work if `Ready` condition is set we don't really care if `ContainerReady` was set first. And if `Ready` is false or not present we just check the `ContainerReady` and we don't care about the timestamps. Right?

SergeyKanzhelev commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,366 @@
+    string podNamespace = 2;
+    string podName = 3;
+    bool static = 4;
+	repeated Conditions conditions = 5;
```
will we be able to say if Pod is terminating?  Do we need this information?

kl52752 replied:
No you are right, I added deletion timestamp to the pod Info to cache terminating pod's.

SergeyKanzhelev commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,366 @@
+}
+
+// PodInfo returns a Pod details and list of status Conditions
+message PodInfo {
```
will PodInfo has a `Ready bool`?

SergeyKanzhelev commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,366 @@
+    rpc GetPodConditions(GetPodConditionsRequest) returns (PodInfo) {}
+}
+
+// Conditions aligns with v1.PodCondition
```
let's keep the naming than

SergeyKanzhelev replied:
Also copying the PodConditionType constants here would make sense

SergeyKanzhelev replied:
One thing we can have as a beta or GA graduation criteria is some kind of test that will ensure that changes in PodCondition are reflected in this API

SergeyKanzhelev reviewed (commented):
Generally lgtm, aligned with the goals, and key decisions. Left some comments and questions on specific APIs and a naming. But rather than that it is lgtm for me

robscott commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,366 @@
+
+We are proposing to create a new kubelet API that will return pod's readiness information.
+* The API will return data about both Static and Regular Pods.
+* The API will not return partial data. If kubelet does not know actual information about workloads then appropriate error code will be returned.
```
I don't think we ever want to redirect to API Server. If clients want that information they can make that request separately. As far as the specific error code, it probably depends on if gRPC or HTTP are chosen.

kl52752 replied:
the gRPC error code FAILED_PRECONDITION will be returned if kubelet is in init phase and does not have all information about the workloads.

robscott commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,366 @@
+| Impact | Description|
+| ------- | ------------ |
+| + | Reliability - for various reasons kube-API might not be available but this doesn’t mean that local workloads are not accessible and on node system workloads should have the most recent data about pod's readiness even when kube-API is unreachable. |
+| + | Scalability - Reduce the load on kube-API be reducing the number of watchers and using kubelet to fetch local pods data. Fetching only pods limited to one node is costly operation for kube-API. |
```
Updated, thanks!

kl52752 commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,366 @@
+
+### Goals
+
+The goal of this API is to expose Pod readiness information directly from the source - kubelet, independent on Control Plane availability. This would remove the need for node-local components to request this node-local information from the Kubernetes API Server. Kubelet already has a podresources endpoint ([2403-pod-resources-allocatable-resources](https://github.com/kubernetes/enhancements/tree/ ...
```
fixed. By readiness information I mean the conditions that are computed by the kubelet. Based on those information the client can understand the pod's status and if it is ready to serve the traffic.

kl52752 commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,366 @@
+| ------- | ------------ |
+| + | Reliability - for various reasons kube-API might not be available but this doesn’t mean that local workloads are not accessible and on node system workloads should have the most recent data about pod's readiness even when kube-API is unreachable. |
+| + | Scalability - Reduce the load on kube-API be reducing the number of watchers and using kubelet to fetch local pods data. Fetching only pods limited to one node is costly operation for kube-API. |
+| + | Safety - Read-only API will not add security risks. |
```
The only information that this API exposes about the pod is its readiness and based on that data, the client can see if the pod is scheduled, running or ready to serve the traffic. We believe  that those are the information that every workload running on the node can see and they are not sensitive or could be misused. This API is also read only.

soltysh commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,447 @@
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
```
Nit: make sure to tag the `(R)` marked points

soltysh commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,447 @@
+## Summary
+
+Proposal to add a new Kubelet gRPC API with endpoint returning local Pods readiness information.
+Serving Pods statuses by Kubelet within a Node will increase reliability and reduce load to the Kubernetes API Server and traffic outside the node. A connectivity issue between Node and Control Plane should not impact workloads which depend on Pods readiness statuses.
```
```suggestion
Serving that information by Kubelet within a Node will increase reliability and reduce load to the Kubernetes API Server and traffic outside the node. A connectivity issue between Node and Control Plane should not impact workloads which depend on Pods readiness statuses.
```
Just to be on the safe side wrt the scope :wink:

kl52752 replied:
done

soltysh commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,447 @@
+
+Proposal to add a new Kubelet gRPC API with endpoint returning local Pods readiness information.
+Serving Pods statuses by Kubelet within a Node will increase reliability and reduce load to the Kubernetes API Server and traffic outside the node. A connectivity issue between Node and Control Plane should not impact workloads which depend on Pods readiness statuses.
+This API is open to future extension but added information should be limited to pod's readiness information.
```
This belongs in the future work section at the end of the document. Summary should be strictly about the current work.

kl52752 replied:
done

soltysh commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,447 @@
+| + | Safety - Read-only API will not add security risks. |
+| + | Reduce resource consumption by workload. Using Kube-API we can fetch objects like PodSpec or PodStatus, for some on-node workload this might be unnecessary, with this API workload can reduce the resource consumption. |
+| - | This API will add load to Kubelet (mitigation: API will be rate limited) |
+| - | Kubelet does not support RBAC authorization for gRPC. (mitigation: This API is designed to be accessible for all workloads running on the node without the authorization. The unix socket will be used for the connection and all exposed data will be carefully reviewed.) |
```
Please add here, that because you don't support RBAC you are limiting the scope of data to just readiness.

kl52752 replied:
done

soltysh commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,447 @@
+source - Kubelet, independent of Control Plane availability by returning kubelet
+owned conditions that are set for pod. This would remove the need for node-local
+components to request this node-local information from
+the Kubernetes API Server.
```
My suggestion would be to change the first sentence slightly:

The goal of this API is to expose Pod readiness information directly from the
source - Kubelet, independent of Control Plane availability ~~by returning kubelet
owned conditions that are set for pod~~. This would remove the need for node-local
components to request this node-local information from
the Kubernetes API Server.

I think that proposed removed part doesn't provide the information needed, and again is confusing wrt conditions.

kl52752 replied:
done

soltysh commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,447 @@
+### What kind of API to chose?
+
+Our goal is to expose the API locally on the node and without Control Plane
+dependency. That mean that we want to avoid authorization which will be done via
```
```suggestion
dependency. That means that we want to avoid authorization which will be done via
```

kl52752 replied:
done

soltysh commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,447 @@
+| Misuse of the API  | High |  This API is Read-only. We will expose only a small portion of the pod's information related to the pod readiness. Exposed data does not contain sensitive information that could be used in a malformed way. |
+| Unauthorized access to the API | Medium | This API is designed to be accessed by all on-node workloads. There is no need to authorize the clients to access this API. |
+| Kubelet restart [issue](https://github.com/kubernetes/kubernetes/issues/100277) | High | This API should serve only complete information about workloads readiness. If Kubelet is in the init phase and not all pod’s readiness information is known, then the API should report the error. |
+| Kube-API is down or unreachable | Low | Kube-API availability should not impact this API. When the control plane is down or unreachable but Kubelet is working properly this API should return most recent data about local Pods readiness that are computed by Kubelet even if those data were not reported or accepted by Kube-API. |
```
I'd like to add one more risk associated with exposing all node's pods readiness:
- enabling listing all pods on a given node (through their readiness) can expose information about what workloads are running in the node/cluster. The API rate limiting mentioned above is one mitigation, but very likely other might need to be added too.

kl52752 replied:
done

soltysh commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,447 @@
+example, networking components may unnecessarily route to unhealthy endpoints
+if they're relying on stale information from the API Server.
+
+The new API should return Pods statuses that are computed by Kubelet even if
```
```suggestion
The new API should return Pods readiness that are computed by Kubelet even if
```

soltysh commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,447 @@
+
+- [ ] Feature implemented behind a feature flag.
+- [ ] e2e tests completed and enabled.
+- [ ] sig-auth input on proposed lack of authorization for this API.
```
:+1: - thank you!

soltysh commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,447 @@
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+Not Applicable.
```
You will need to at least manually test this path prior to beta graduation.

soltysh commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,447 @@
+
+###### Will enabling / using this feature result in introducing new API types?
+
+N/A
```
You are exposing new API, even though it's only node-local.

kl52752 replied:
done

soltysh reviewed (commented):
A few minor comments, but I think this is good for @johnbelamaric to have a look at

soltysh commented:
from prr-shadow pov, this looks good
/assign @johnbelamaric

dchen1107 commented:
/lgtm
/approve 

There are two unresolved nits, and please address them here or with another pr.

SergeyKanzhelev commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,451 @@
+contain information about pod readiness status.
+
+Kubelet is responsible for computing Pod status and stores it in a local cache.
+We want to create a new ghRPC API that will expose pod conditions that are computed by
```
```suggestion
We want to create a new gRPC API that will expose pod conditions that are computed by
```

SergeyKanzhelev commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,451 @@
+
+### Proposed API
+
+We propose to add new gPRC API `status` in Kubelet, listening on a unix socket at `/var/lib/Kubelet/status/Kubelet.sock`. The endpoint will be versioned.
```
a note here - with podresources we FORCE it to be a subfolder of kubelet root. We need to make sure it is not the case here.

SergeyKanzhelev replied:
some other thinks to spec out - whether it can be turned off: see some discussions here: https://github.com/kubernetes/kubernetes/issues/120625

It can be a goal for beta

SergeyKanzhelev commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,451 @@
+
+  - Feature gate enable/disable tests.
+  - Test API work when Kubelet restarts.
+  - Test API work when Control Plane is unreachable.
```
let's also add a test for the kubelet in Standalone mode

kl52752 replied:
done

SergeyKanzhelev commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,451 @@
+
+// PodCondition aligns with v1.PodCondition.
+message PodCondition {
+    ConditionType Type = 1;
```
let's keep as much as possible in sync with types.go. Lets call this one PodConditionType

SergeyKanzhelev commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,451 @@
+	DisruptionTarget = 4;
+}
+
+// ConditionStatus aligns with v1.ConditionStatus
```
it is string in `v1.ConditionStatus`.

SergeyKanzhelev replied:
please make a note about this differnce in the type comment. I don't see much harm here to switch to strongly typed enum

SergeyKanzhelev commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,451 @@
+
+## Design Details
+
+### Proposed API
```
there are still some comments here. I want to make sure we are as aligned with types we have as possible so if we need to bring more fields and types - it is easy to do.

SergeyKanzhelev replied:
I wouldn't block the KEP on this, some of it can be aligned during the PR review.

SergeyKanzhelev replied:
cc @mrunalp to be on the same page on not blocking on this.

johnbelamaric commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,449 @@
+* The API should return the most recent information about Pods computed by the
+  Kubelet even when those data were not reported or accepted by kube-API.
+* The API will be read-only and accessible for on-node workloads (we will use a
+  unix socket for the connection) without authorization.
```
..beyond standard unix permissions.

(non-blocking nit)

kl52752 replied:
done

johnbelamaric commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,449 @@
+
+Our goal is to expose the API locally on the node and without Control Plane
+dependency. That means that we want to avoid authorization which will be done via
+kube-API. To fullfil this requirement we decided to use unix socket to access
```
s/fullfil/fulfill/

kl52752 replied:
done

johnbelamaric commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,449 @@
+| Risk                                                      | Impact        | Mitigation |
+| --------------------------------------------------------- | ------------- | ---------- |
+| Too many requests to the API impacting the Kubelet performances | High          | Rate limiting the API. |
+| Misuse of the API  | High |  This API is Read-only. We will expose only a small portion of the pod's information related to the pod readiness. Exposed data does not contain sensitive information that could be used in a malformed way. |
```
s/malformed/malicious/

This is a little strong to say it *cannot* be used this way, but if SIG security is OK with this I wouldn't argue.

I could imagine if there were some local service that enforced some policy, knowing that it is not yet ready could be useful to an attacker.

(non-blocking comment)

kl52752 replied:
done

johnbelamaric commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,449 @@
+| Too many requests to the API impacting the Kubelet performances | High          | Rate limiting the API. |
+| Misuse of the API  | High |  This API is Read-only. We will expose only a small portion of the pod's information related to the pod readiness. Exposed data does not contain sensitive information that could be used in a malformed way. |
+| Kubelet restart [issue](https://github.com/kubernetes/kubernetes/issues/100277) | High | This API should serve only complete information about workloads readiness. If Kubelet is in the init phase and not all pod’s readiness information is known, then the API should report the error. |
+| Unauthorized access to the API | Medium | This API is designed to be accessed by all on-node workloads. There is no need to authorize the clients to access this API. |
```
You can use unix permissions to control access to the socket.

kl52752 replied:
done

johnbelamaric commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,449 @@
+
+###### How can an operator determine if the feature is in use by workloads?
+
+Look at the `pod_info_endpoint_requests_list` and `pod_resources_info_requests_get` metric exposed by the Kubelet.
```
are these the right metrics? did you mean `pod_info` not `pod_resources_info`? Should it be `pod_readiness` not `pod_info`...info seems very generic?

kl52752 replied:
fixed

johnbelamaric reviewed (changes requested):
Some minor changes requested but otherwise LGTM for PRR.

wojtek-t commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,450 @@
+
+###### Are there any tests for feature enablement/disablement?
+
+e2e test will demonstrate that when the feature gate is disabled, the API returns the appropriate error code.
```
This will effectively be a feature test.

Given this is purely in-memory feature - enablement/disablement is effectively equal to the feature test though [but let's clarify that in the text here].

SergeyKanzhelev commented on keps/sig-node/4188-kubelet-pod-readiness-api/kep.yaml:
```diff
@@ -0,0 +1,43 @@
+participating-sigs: []
+status: implementable
+creation-date: 2023-09-05
+reviewers:
```
please specify people explicitly

robscott replied:
Updated, thanks!

dchen1107 commented on keps/sig-node/4188-kubelet-pod-readiness-api/kep.yaml:
```diff
@@ -0,0 +1,43 @@
+creation-date: 2023-09-05
+reviewers:
+approvers:
+  - "@sig-node-tech-leads"
```
I am the approver here.

robscott replied:
Updated, thanks!

SergeyKanzhelev commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
````diff
@@ -0,0 +1,450 @@
+```protobuf
+service Status {
+    // ListPodConditions returns a of List of PodConditions
+    rpc ListPodConditions(Empty) returns (PodConditionsListResponse) {}
````
let's have a request type so we can add filters later if needed

SergeyKanzhelev replied:
(instead of Empty have a `PodConditionsListRequest` which is currently empty

robscott replied:
Updated, thanks!

SergeyKanzhelev commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,450 @@
+// PodConditionsResponse returns a stream of List of PodConditionsResponse
+// Whenever a Pod state changes it will return the new list.
+message PodConditionsListResponse {
+	repeated PodReadiness Pods = 1;
```
should we name it `PodStatus` to make it more generic if we want to expose more status information in future? PodReadiness is too specific for the single purpose.

robscott replied:
Updated, thanks!

SergeyKanzhelev commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,450 @@
+    rpc ListPodConditions(PodConditionsListRequest) returns (PodConditionsListResponse) {}
+    // WatchPodConditions returns a stream of List of PodConditions
+    // Whenever a pod state change api returns the new list
+    rpc WatchPodConditions(Empty) returns (stream PodConditionsListResponse) {}
```
how will this be imlpemented? Are we returning individual events when something is changed? Or return all Pods from time to time? Should we have an events-based response instead of the same response type?

SergeyKanzhelev commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,450 @@
+    string podNamespace = 2;
+    string podName = 3;
+    bool static = 4;
+	repeated PodCondition conditions = 5;
```
please fix the indentation in this section. Hard to read

robscott replied:
Updated, thanks!

SergeyKanzhelev commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,450 @@
+    // Whenever a pod state change api returns the new list
+    rpc WatchPodConditions(Empty) returns (stream PodConditionsListResponse) {}
+    // GetPodConditions returns a PodReadiness for given pod's UID
+    rpc GetPodConditions(PodConditionsRequest) returns (PodReadiness) {}
```
this method type is confusing. Method says the Pod Condition will be returned and Pod Readiness will be returned.

I think the best would be to introduce the PodStatus aligned with https://github.com/kubernetes/kubernetes/blob/f19b62fc0914b38941922afefd1e34eb55f87ee7/pkg/apis/core/types.go#L3687 and rename all methods to request PodStatus

robscott replied:
Updated, thanks!

johnbelamaric commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,450 @@
+
+###### What specific metrics should inform a rollback?
+
+`pod_rediness_endpoint_errors_get` - but only with feature gate `PodReadinessAPI` enabled.
```
s/rediness/readiness/

(non-blocking nit)

robscott replied:
Updated, thanks!

johnbelamaric commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,450 @@
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+- [X] Metrics
+  - Metric name:  `pod_info_endpoint_requests_total`, `pod_info_endpoint_requests_list` and `pod_info_endpoint_requests_get`.
```
update names here

robscott replied:
Updated, thanks!

johnbelamaric commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,450 @@
+
+###### Are there any missing metrics that would be useful to have to improve observability of this feature?
+
+As part of this feature enhancement, per-API-endpoint resources metrics are being added; `pod_info_endpoint_requests_get`, `pod_info_endpoint_requests_list` add `pod_info_endpoint_errors_get`.
```
and here

johnbelamaric commented on keps/sig-node/4188-kubelet-pod-readiness-api/kep.yaml:
```diff
@@ -0,0 +1,46 @@
+
+# The following PRR answers are required at beta release
+metrics:
+  - pod_info_endpoint_requests_total
```
and here

robscott replied:
Updated, thanks!

johnbelamaric reviewed (commented):
some nits

robscott commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,450 @@
+    rpc ListPodConditions(PodConditionsListRequest) returns (PodConditionsListResponse) {}
+    // WatchPodConditions returns a stream of List of PodConditions
+    // Whenever a pod state change api returns the new list
+    rpc WatchPodConditions(Empty) returns (stream PodConditionsListResponse) {}
```
Updated, thanks!

johnbelamaric commented:
/approve

But may be nice to fix the nits before someone does lgtm

SergeyKanzhelev commented on keps/sig-node/4188-kubelet-pod-readiness-api/README.md:
```diff
@@ -0,0 +1,453 @@
+    rpc ListPodStatus(PodStatusListRequest) returns (PodStatusListResponse) {}
+    // WatchPodStatus returns a stream of List of PodStatus
+    // Whenever a pod state change api returns the new list
+    rpc WatchPodStatus(PodStatusWatchRequest) returns (stream PodStatusWatchResponse) {}
```
returning individual PodStatus when it has changed would be the best design. This way there is no overhead of returning everything every time a single Pod has changed

SergeyKanzhelev reviewed (approved):
/lgtm

1. I still have questions to the API design, but those can be ironed at implementation. Overall the KEP lgtm. especially around Watch API.
2. This comment needs to be resolved by beta: https://github.com/kubernetes/enhancements/pull/4184#discussion_r1345044999 Please add it as a goal for beta.

Other than that, great KEP and I am happy we will have this API!

deads2k commented:
I re-found this while having a look at https://github.com/kubernetes/enhancements/pull/4615/files and noticing that it was also dealing with the power and limitations of the existing grpc kubelet endpoint.  In particular, I noticed that
1. there is more than one resource needed for read powers from the kubelet
2. even with a single resource, partitioning between the clients is desired (authentication)
3. the clients using the existing grpc extension are trying tunnel http over the grpc to be able to re-use the existing client and controller libraries.  Even the APIs don't match exactly, the dynamic client and controller libraries still operate extremely well.
4. versioning of APIs is important for adoption and to allow old and new extensions to coexist

As I recall, versioning was well handled by this design, but the trajectory of resource grpc endpoint could inform future kubelet serving extensions.
